Account for decoy items in score

arbrandes opened the pull request:
## Description

In order to take the correct non-placement of decoy items into account, this PR implements the student score as `score = ((correctly placed items) + (correctly unplaced decoys)) / (non-decoys + decoys)`

## Dependencies

- #87 (now merged)

## JIRA

- [SOL-1988](https://openedx.atlassian.net/browse/SOL-1988)

## Sandbox URLs

- [LMS](http://lms.arbrand.es/courses/course-v1:test+tt101+run/courseware/chapter/drag_and_drop_v2/)
- [Studio](http://cms.arbrand.es/container/block-v1:test+tt101+run+type@vertical+block@drag_and_drop_v2)

## Latest commit

- a1b6931 *(updated 2016-08-18)*

## Testing instructions

1. Go to the [pre-deployed DnDv2 instance](http://lms.arbrand.es/courses/course-v1:test+tt101+run/courseware/chapter/drag_and_drop_v2/) on the LMS, and log in as `staff@example.com`.
2. Click on `Staff Debug Info`, and `Delete Student State` for the `staff@example.com` user.
3. Close the debug info window, drag all items into their correct places, and finally drag the `I don't belong anywhere` item anywhere onto the image.
4. Click `Submit`.
5. Go to the [progress page](http://lms.arbrand.es/courses/course-v1:test+tt101+run/progress), and check that the score for `Drag and Drop v2` is **80%**, or 4 out of 5.

## Screenshots

### Suggested submission

![screenshot](https://cloud.githubusercontent.com/assets/759355/17751446/85f7b13c-649d-11e6-9ef9-fcb78f8922ea.png)

### Expected grade

![screenshot2](https://cloud.githubusercontent.com/assets/759355/17751445/85f792ec-649d-11e6-8038-df15e290f862.png)

e-kolpakov commented on drag_and_drop_v2/drag_and_drop_v2.py:
```diff
@@ -516,11 +516,14 @@ def _get_item_stats(self):
+        decoy_items = [str(item['id']) for item in all_items if self._get_item_zones(item['id']) == []]
+        decoy_items_placed = [item for item in decoy_items if item in item_state]
+
+        required_count = len(required_items) + len(decoy_items)
```
@arbrandes this could be simplified to just `len(all_items)`

e-kolpakov commented on drag_and_drop_v2/drag_and_drop_v2.py:
```diff
@@ -516,11 +516,14 @@ def _get_item_stats(self):
+        decoy_items_placed = [item for item in decoy_items if item in item_state]
+
+        required_count = len(required_items) + len(decoy_items)
+        correct_count = len(required_items_correct) + len(decoy_items) - len(decoy_items_placed)
```
@arbrandes nit: the way it written right now is a bit confusing - I even started writing "it wouldn't work like expected ...", got to the part where I'm providing an example and than realized that it actually will.

To avoid that kind of confusion, could you please update it to read more like in SOL-1988:  [(# of non-decoy items placed onto a correct zone) + (# of decoy items correctly left unplaced)] ÷ (total # of items), i.e

    decoy_items_in_bank = [item for item in decoy_items if item **not** in item_state]

    correct_count = len(required_items_correct) + len(decoy_items_in_bank)

e-kolpakov commented:
@arbrandes Mostly looks good, but I've got two concerns

* I'm not convinced the test suite contains tests that actually check the new rule in case when decoy items are placed on the board. Could you please add some tests for it (I would say a couple of unit tests to cover all edge cases and one-two integration tests).
* I've significantly [reworked this part of code](https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/87/files#diff-81cf2d60a06d8f43d0df802d9aad9417R751). Depending on which PR lands first, someone will have to resolve conflicts. The conflict should be relatively small, but the change is not that huge too :) On the other hand, making this PR depend on #87 does not seem like a good idea too (I've got a couple of such PRs and it's not playing out well).

So, I would love to see my first concern addressed (and won't approve this PR without it :)). As of the second concern - building this PR on top of #87 will likely save us some merge conflicts, but it might complicate the review process and PR maintenance in general. So I'm leaving it for you to decide.

Also, we'll need a sandbox for this PR - it is required for upstream review, but it is useful for "internal" reviews too. If you have registered for [Opencraft IM](https://console.opencraft.com/) beta test you could build a sandbox yourself. Otherwise, I'll have to do that, but I'm already juggling 4 DnDv2 sandboxes, so I would be grateful to not add one more to the list.

arbrandes commented:
A fair assesment!  I'm going to address them later on today and update the
PR.

I'll also see what I can do for a sandbox on Opencraft IM, or elsewhere.
What do you guys usually use?  That?  Amazon or OpenStack VMs?  Your own
boxes?

As for #87, I suggest we cross that bridge when we get there.  This way,
this PR can be judged on its own merit.  I'll be glad to resolve any
conflicts once the time comes.

Thanks Eugeny! :)

e-kolpakov commented:
@arbrandes for sandbox we just use Opencraft IM and let it build the sandbox. Under the hood I believe it runs on OpenStack, but if you're not using Opencraft IM (you are not oblidged to) you can use whatever you want. :)

arbrandes commented:
@e-kolpakov, reality check: is it even *possible* to drag a decoy item into a zone in assessment mode, in latest master?  I couldn't do it.  (And it would also explain why I can't get even a unit test to do it either).  And if it's not possible, is it part of SOL-1988 to make it so?

Related question: I can't properly edit the dndv2 block from master in Studio due to a handlebar exception.  Is this a known issue?  (Note: I don't have access to the Atlassian issues, for some reason, so I can't check for myself.)

e-kolpakov commented:
@arbrandes Decoy item: no, it is not possible in latest master as assessment mode handling is not yet implemented there. But it is implemented in #87, so it's not part of SOL-1988. But you won't be able to properly  test SOL-1988 without it, so looks like you should rebase on top of #87 after all.

Studio error: No, I don't think we saw that kind of error yet - could you plese post screenshot and/or traceback?

arbrandes commented:
@e-kolpakov, ok, thanks, I'll rebase on #87 and see how it goes there. :)

As for the Studio error, I'll retest on the latest edx-platform master to see if it's still there, and report back.

e-kolpakov commented:
@arbrandes #87 just landed on master, so you've got to rebase now.

Aso, I'm away starting this Friday - so it would be great if you could make it ready for review today or tomorrow, to have some time for review and addressing review comments. Note my Friday is like late Thursday yuor timezone.

arbrandes commented:
@e-kolpakov, roger that, I'll do my best to finish it up today, or failing that, early during the day tomorrow.  Thanks for the heads-up!

arbrandes commented:
@e-kolpakov, rebased on master, conflicts solved, and tests added.  A sandbox, based on the latest edX master (as of yesterday, at least), here:

http://lms.arbrand.es/courses/course-v1:test+tt101+run/courseware/chapter/drag_and_drop_v2/

I'll see about improving connection speed to the sandbox in the future, but this is the best I can do for now.

arbrandes commented:
@e-kolpakov, it would probably be nice to update opencraft/master to the latest edx-solutions/master as well.

e-kolpakov commented:
@arbrandes I've just updated opencraft master to solutions master and it looks like something is wrong with your branch - this PR still shows 5 commits from me. Looks like you've got to rebase it again (make sure to skip my commits :))

e-kolpakov commented on drag_and_drop_v2/drag_and_drop_v2.py:
```diff
@@ -764,7 +770,10 @@ def _get_item_raw_stats(self):
+        decoy_items = set(item_id for item_id in all_items if self._get_item_zones(int(item_id)) == [])
+        decoy_items_unplaced = set(item for item in decoy_items if item not in item_state)
+
+        return required_items, placed_items, correct_items, decoy_items, decoy_items_unplaced
```
@arbrandes it might make sense to convert this 5-tuple to a namedtuple, to avoid unpacking and discarding some values ( `required_ids, placed_ids, correct_ids, __, __`)

arbrandes replied:
Good idea.  Let me try my hand at it.

arbrandes commented:
@e-kolpakov, not sure what the problem is.  I rebased on 7b5bc7d, which is now both edx-solutions and open-craft master.  (It wasn't before, though, so maybe that's what happened.)

e-kolpakov commented on tests/integration/test_interaction.py:
```diff
@@ -674,6 +674,44 @@ def test_do_attempt_feedback_is_updated(self):
         self.assertEqual(self._get_feedback().text, expected_feedback)
 
 
+class AssessmentGradingTest(
```
@arbrandes any reason this is a new class rather than added to existing `AssessmentInteractionTest`?

arbrandes replied:
@e-kolpakov, none other than I originally expected to have more tests.  Turns out it wasn't really necessary, so you're right, a class just for a single test seems overkill.

e-kolpakov commented on tests/integration/test_interaction.py:
```diff
@@ -674,6 +674,44 @@ def test_do_attempt_feedback_is_updated(self):
+
+        # Find the grade event
+        published_grade = None
+        for call in self.publish.call_args_list:
```
@arbrandes suggestion:

    grade_event = next(publish_call_args[0] for publish call in self.publish.call_args_list if publish_call_args[0][1] == 'grade')
    self.assertEqual(grade_event[2], expected_grade)

Not tested, but I hope you see what I mean here.

Alternatively, it might make sense to use `assertIn`:

    expected_event = (self.block, 'grade', expected_grade)
    self.assertIn(expected_event, self.publish_call_arg_list)

arbrandes replied:
Ok, that'll save a couple of lines.  Trying it, thanks!

arbrandes commented:
@e-kolpakov, I addressed your comments and updated the PR.

Note: `assertIn()` didn't work because `self.block` isn't available and `mock.ANY` just gave segfaults.  Go figure.

e-kolpakov commented on drag_and_drop_v2/drag_and_drop_v2.py:
```diff
@@ -741,30 +742,39 @@ def _get_zone_by_uid(self, uid):
     def _get_item_raw_stats(self):
         """
-        Returns a 3-tuple containing required, placed and correct items.
+        Returns a named tuple containing required, decoy, placed, correct
```
@arbrandes nit: correctly placed

e-kolpakov commented on drag_and_drop_v2/drag_and_drop_v2.py:
```diff
@@ -741,30 +742,39 @@ def _get_zone_by_uid(self, uid):
         """
-        Returns a 3-tuple containing required, placed and correct items.
+        Returns a named tuple containing required, decoy, placed, correct
+        placed, and correct decoy items.
```
@arbrandes nit: `correct decoy` feels a bit weird - maybe `decoy items in bank`?

e-kolpakov commented on drag_and_drop_v2/drag_and_drop_v2.py:
```diff
@@ -741,30 +742,39 @@ def _get_zone_by_uid(self, uid):
+        required = set(item_id for item_id in all_items if self._get_item_zones(int(item_id)) != [])
+        placed = set(item_id for item_id in all_items if item_id in item_state)
+        correct_placed = set(item_id for item_id in placed if item_state[item_id]['correct'])
+        decoy = set(item_id for item_id in all_items if self._get_item_zones(int(item_id)) == [])
```
@arbrandes nit: convert `all_items` to set, than `decoy = all_items - required_items`

e-kolpakov commented on drag_and_drop_v2/drag_and_drop_v2.py:
```diff
@@ -741,30 +742,39 @@ def _get_zone_by_uid(self, uid):
+        correct_decoy = set(item for item in decoy if item not in item_state)
 
-        return required_items, placed_items, correct_items
+        Items = namedtuple('Items', 'required, placed, correct_placed, decoy, correct_decoy')
```
@arbrandes I've never seen namedtuple *defined* within a method, but maybe I'm just a bit conservative. Since namedtuple is much like a lightweight class, they are usually defined in the module code, like classes do (pylint complains about the names though, but we usually suppress it, since it does not understand the intention right).

So, could you please move `Items` namedtuple declaration into `utils.py`, and probably rename it to something more intuitive, i.e. ItemStats?

e-kolpakov commented:
@arbrandes some feedback for PR description in general - it would be very helpful if you could include explicit testing instructions, link to a sandbox and other relevant info in PR description. I'm not sure who will be sending PR against solutions master (you, me or someone else from Opencraft), but upstream will expect PR description to contain at least that information, and probably more.

In case you will be sending a PR, you could use any recent DnDv2 PR (i.e. edx-solutions#87, edx-solutions#92, edx-solutions#93) as an example and/or template.

e-kolpakov commented:
@arbrandes one more thing :) Sandbox backend code and frontend code are inconsistent: 

![image](https://cloud.githubusercontent.com/assets/3330048/17738977/23bbbc26-649c-11e6-8d84-89705b68a0f3.png)

These [Object object] should be overall feedback messages - the format have changed from plain string to `{"message": ..., "message_class": ...}`. I believe updating assets on the sandbox would be enough (`paver update_assets`), but you might need to restart workers as well.

arbrandes commented:
@e-kolpakov, nits all picked, and sandbox updated properly. ;)

I'll get to work on the PR description.

arbrandes commented:
@e-kolpakov, description updated.

e-kolpakov commented:
:+1:

- [x] I tested this: 
  - Placing all required items and leaving all decoys in item bank gives 1.0 score.
  - Placing all required items and all/some decoys reduce the score by (count_of_placed_decoys) / (total items)
  - Placing only decoys gives 0 score
- [x] I read through the code
- [ ] Includes documentation

e-kolpakov commented:
@arbrandes ^^

Just in case conditional +1 means you should still address some problems outlined, but I trust you to fix it without me reviewing it again.

bradenmacdonald commented:
Please open a PR against the main edx-solutions repository. Don't merge into the OpenCraft master. Copy your +1 to the new PR once it's open if you're still available.

e-kolpakov commented:
@bradenmacdonald 

> Please open a PR against the main edx-solutions repository. 

I assume this part is for @arbrandes 

> Copy your +1 to the new PR once it's open if you're still available.

And this one is for me?

Otherwise it would be a bit funny to post a PR and case a +1 on it :)

bradenmacdonald commented:
Yes, sorry for not being clear.

arbrandes commented:
@e-kolpakov, here's the PR: https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/96

bradenmacdonald commented:
Thanks! Actually in the future it might be better for someone from OpenCraft to open the PR, because otherwise the OSPR bot creates an OSPR ticket in JIRA, which we don't want.
